Serialization: Use a DeclVisitor to implement deserialization safety

tshortli opened the pull request:
Refactor deserialization safety to use a `DeclVisitor` CRTP instead of ad-hoc casts. This ensures every kind of decl is handled explicitly.
    
Resolves rdar://115456536

tshortli committed:
Serialization: Use a DeclVisitor to implement deserialization safety.

Refactor deserialization safety to use a `DeclVisitor` CRTP instead of ad-hoc
casts. This ensures every kind of decl is handled explicitly.

Resolves rdar://115456536

tshortli committed:
Serialization: Handle enum cases in ExternallyAccessibleDeclVisitor.

tshortli committed:
Tests: Add a typealias test case to lazy_typecheck.swift.

tshortli committed:
Serialization: Handle IfConfigDecls in ExternallyAccessibleDeclVisitor.

tshortli committed:
Serialization: Handle precedence groups and operators in ExternallyAccessibleDeclVisitor.

tshortli commented:
@swift-ci please smoke test

xymus commented on lib/Serialization/Serialization.cpp:
```diff
@@ -3333,44 +3351,97 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
-      //        storage for them to preserve the same safeness.
+      //        storage for them to preserve the same accessibility.
       if (var->isLazyStorageProperty())
         return true;
```
As the fixme says, this is something that likely marks too many things as safe/accessible. We should keep an eye on it.

xymus commented on lib/Serialization/Serialization.cpp:
```diff
@@ -3276,23 +3276,39 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
-  static bool isDeserializationSafe(const Decl *decl) {
-    if (auto ext = dyn_cast<ExtensionDecl>(decl)) {
-      // Consider extensions as safe as their extended type.
+  class ExternallyAccessibleDeclVisitor
```
I don't mind the name but I think we usually use "exported" for the idea of externally accessible. The term may already be overloaded however.

tshortli replied:
Yeah, that's a good point. Maybe I should rename this and follow through on extracting this visitor into it's own file the hope that we can unify some or all of the other exportability checks in the compiler.

xymus replied:
There is a set of `isExported` functions that are also getting very close to this version as you're reducing the difference between was is serialization safe and just what's exported.

tshortli replied:
I pulled the class out into a header and renamed it to `DeclExportabilityVisitor`.

xymus reviewed (approved):
It's a clear improvement, thanks!

tshortli committed:
AST: Expose DeclExportabilityVisitor.

Extract `ExternallyAccessibleDeclVisitor` from `Serialization.cpp` into its own
header and rename it to `DeclExportabilityVisitor` to better align with
terminology in other parts of the compiler. Ideally, `DeclExportabilityVisitor`
should become the canoncial implementation of for exportability checks,
consolidating logic that is currently spread between serialization, module
interface printing, TBDGen, and type checking. For now, it is only used in
serialization to implement serialization safety checks.

tshortli commented:
@swift-ci please smoke test
